python/ql/lib/semmle/python/dataflow/new/internal/VariableCapture.qll

```diff
@@ -0,0 +1,189 @@
+      // covered by global variable handling.
+      this.getScope() = f and
+      this.getAnAccess().getScope() != f
+    }
+
+    Callable getCallable() { result = f }
+
+    Location getLocation() { result = f.getLocation() }
+
+    /** Gets a scope that captures this variable. */
+    Scope getACapturingScope() {
+      result = this.getAnAccess().getScope().getScope*() and
+      result.getScope+() = f
+    }
+  }
+
+  class CapturedParameter extends CapturedVariable {
+    CapturedParameter() { this.isParameter() }
+
+    ControlFlowNode getCfgNode() { result.getNode().(Parameter) = this.getAnAccess() }
+  }
+
+  class Expr extends ExprCfgNode {
+    predicate hasCfgNode(BasicBlock bb, int i) { this = bb.getNode(i) }
+  }
+
+  class VariableWrite extends ControlFlowNode {
+    CapturedVariable v;
+
+    VariableWrite() { this = v.getAStore().getAFlowNode() }
+
+    CapturedVariable getVariable() { result = v }
+
+    predicate hasCfgNode(BasicBlock bb, int i) { this = bb.getNode(i) }
+  }
+
+  class VariableRead extends Expr {
+    CapturedVariable v;
+
+    VariableRead() { this = v.getALoad().getAFlowNode() }
+
+    CapturedVariable getVariable() { result = v }
+  }
+
+  class ClosureExpr extends Expr {
+    ClosureExpr() {
+      this.getNode() instanceof CallableExpr
+      or
+      this.getNode() instanceof Comp
+    }
+
+    predicate hasBody(Callable body) {
+      body = this.getNode().(CallableExpr).getInnerScope()
+      or
+      body = this.getNode().(Comp).getFunction()
+    }
+
+    predicate hasAliasedAccess(Expr f) { closureFlowStep+(this, f) and not closureFlowStep(f, _) }
+  }
+}
+
+class CapturedVariable = CaptureInput::CapturedVariable;
+
+class ClosureExpr = CaptureInput::ClosureExpr;
+
+module Flow = Shared::Flow<Location, CaptureInput>;
+
+private Flow::ClosureNode asClosureNode(Node n) {
+  result = n.(SynthCaptureNode).getSynthesizedCaptureNode()
+  or
+  result.(Flow::ExprNode).getExpr() = n.(CfgNode).getNode()
+  or
+  result.(Flow::VariableWriteSourceNode).getVariableWrite() = n.(CfgNode).getNode()
+  or
+  result.(Flow::ExprPostUpdateNode).getExpr() =
+    n.(PostUpdateNode).getPreUpdateNode().(CfgNode).getNode()
+  or
+  result.(Flow::ParameterNode).getParameter().getCfgNode() = n.(CfgNode).getNode()
+  or
+  result.(Flow::ThisParameterNode).getCallable() =
+    n.(SynthCapturingClosureParameterNode).getCallable()
+}
+
+predicate storeStep(Node nodeFrom, CapturedVariableContent c, Node nodeTo) {
+  Flow::storeStep(asClosureNode(nodeFrom), c.getVariable(), asClosureNode(nodeTo))
+}
+
+predicate readStep(Node nodeFrom, CapturedVariableContent c, Node nodeTo) {
+  Flow::readStep(asClosureNode(nodeFrom), c.getVariable(), asClosureNode(nodeTo))
+}
+
+predicate valueStep(Node nodeFrom, Node nodeTo) {
+  Flow::localFlowStep(asClosureNode(nodeFrom), asClosureNode(nodeTo))
+}
+
+/**
+ * Provides predicates to understand the behaviour of the variable capture
+ * library instantiation on Python code bases.
+ */
```

Comment:
this is super abstract, and doesn't mention anything about the expectation that the `unmapped*` predicates ideally should have 0 results.

Reply:
Done